### DISCUSSION TITLE
Hide specific files and folders

### DISCUSSION THREAD

**Discussion by gruanlc** [drakkan/sftpgo#669] [Q&A] [answered]
Hi, 
is it possible to hide specific files, file types or folders contained in a folder/path?
I have only found a possibility to deny the access of files and folders. But these files and folders are still visible in lists.
Sometimes it would be very helpful, if I could hide specific files/folders for specific user.

**Comment by drakkan** [comment, marked as answer]
Hi,

it is currently not possible. I'm aware that some proprietary software similar to SFTPGo have this feature but it could be very slow. Think if you have a directory with 100.000 files and folders and for each item you have to check a file pattern or such

**Reply by gruanlc** [reply]
Is the already implemented mechanism to "allow/deny" files and folders so completly different?

**Reply by drakkan** [reply]
> Is the already implemented mechanism to "allow/deny" files and folders so completly different?

Yes, because we just check a single path before a download/upload request, not all the paths in a dir, see here:

https://github.com/drakkan/sftpgo/blob/main/dataprovider/user.go#L820

**Reply by drakkan** [reply]
I could also add this in the future but it could cause issues for directories with many files. If you only use SFTPGo and configure the correct file patterns the users are unable to upload the unwanted files. If you upload using other methods or you just proxy another SFTP server you could have unwanted files in the list. Is this your use case?

**Reply by gruanlc** [reply]
I see the problem in case of many files. In my case all could be solved on folder level.

My use case is following:

I have a (complex structured) data archive with several data products (e.g. product1, product2, ...) and different release levels (e.g. alpha, beta, final) and thousands of data files. Now, I want to give user-specific access (read only) to a "part" of this data archive. For exampale, a user1 should have access to "product1" but only release level "beta" and "final". A user2 should have access to all products but only release level "final". And so on.

My idea was, to add user-specific filter for folders, e.g. hide folders which include "beta" (e.g. per regexp ".*beta.*" or file mask "*beta*").

Of course, the access itself can already be solved using the "allow/deny" functionality. But than, all with "deny" marked folders are visible to users. And they would always ask, why can they not access this and this.

Another option would be to create one virtual folder per specific combination (e.g. product1 + beta). But that would lead to a very long list of virtual folders. And a lot of mounting in the user-definitions.

Maybe there are further options which I do not see.

**Reply by drakkan** [reply]
> I see the problem in case of many files. In my case all could be solved on folder level.
> 
> My use case is following:
> 
> I have a (complex structured) data archive with several data products (e.g. product1, product2, ...) and different release levels (e.g. alpha, beta, final) and thousands of data files. Now, I want to give user-specific access (read only) to a "part" of this data archive. For exampale, a user1 should have access to "product1" but only release level "beta" and "final". A user2 should have access to all products but only release level "final". And so on.
> 
> My idea was, to add user-specific filter for folders, e.g. hide folders which include "beta" (e.g. per regexp "._beta._" or file mask "_beta_").
> 
> Of course, the access itself can already be solved using the "allow/deny" functionality. But than, all with "deny" marked folders are visible to users. And they would always ask, why can they not access this and this.
> 
> Another option would be to create one virtual folder per specific combination (e.g. product1 + beta). But that would lead to a very long list of virtual folders. And a lot of mounting in the user-definitions.
> 
> Maybe there are further options which I do not see.

The use case is valid, please give me a few days to think about it a little more, thanks

**Comment by drakkan** [comment]
> Hi, is it possible to hide specific files, file types or folders contained in a folder/path? I have only found a possibility to deny the access of files and folders. But these files and folders are still visible in lists. Sometimes it would be very helpful, if I could hide specific files/folders for specific user.

Hi,

this is now supported: 

https://github.com/drakkan/sftpgo/commit/c3831de94ef73fc9f453520c448c4962b1c5a6a6
https://github.com/drakkan/sftpgo/commit/a6ed6fc721848602409f3e6f24b6007a2a5fbe6e

Please let me know how it works for your real use case, there are many things that can go wrong. A feedback/accurate testing is really appreciated.
